src/middleware/devtools.ts

```diff
@@ -61,107 +74,190 @@ export const devtools =
   (
     set: CustomSetState,
     get: CustomGetState,
-    api: CustomStoreApi & StoreApiWithDevtools<S> & { dispatch?: unknown }
+    api: CustomStoreApi &
+      StoreApiWithDevtools<S> & {
+        dispatch?: unknown
+        dispatchFromDevtools?: boolean
+      }
   ): S => {
-    let extension
-    try {
-      extension =
-        (window as any).__REDUX_DEVTOOLS_EXTENSION__ ||
-        (window as any).top.__REDUX_DEVTOOLS_EXTENSION__
-    } catch {}
-
-    if (!extension) {
+    let devtoolsOptions =
+      options === undefined
+        ? { name: undefined }
+        : typeof options === 'string'
+        ? { name: options }
+        : options
+
+    let extensionConnector =
+      (window as any).__REDUX_DEVTOOLS_EXTENSION__ ??
+      (window as any).top.__REDUX_DEVTOOLS_EXTENSION__
```

Comment:
I didn't now `top` is not nullable. 🤯 

I feel like we should fix devtools in jotai and valtio too. cc @Aslemammad